src/Command/ParamAwareCommandTrait.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/**
+ * @see       https://github.com/laminas/laminas-cli for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-cli/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-cli/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Laminas\Cli\Command;
+
+use Laminas\Cli\Input\InputParamInterface;
+use Laminas\Cli\Input\NonHintedParamAwareInput;
+use Laminas\Cli\Input\TypeHintedParamAwareInput;
+use PackageVersions\Versions;
+use RuntimeException;
+use Symfony\Component\Console\Helper\HelperSet;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+use function is_array;
+use function sprintf;
+use function str_replace;
+use function strstr;
+
+/**
+ * Provide promptable input options to your command.
+ *
+ * Compose this trait in your symfony/console command in order to provide the
+ * ability to define options that will result in interactive prompts when not
+ * provided. Such parameters can be added using the `addParam()` construct. When
+ * present, you can then use `$input->getParam($name)` to retrieve the value. If
+ * the value was provided as an option, that value will be returned; otherwise,
+ * it will prompt the user for the value.
+ */
+trait ParamAwareCommandTrait
+{
+    /** @var array array<string, InputParamInterface> */
+    private $inputParams = [];
+
+    /**
+     * @return $this
+     * @throws RuntimeException
+     */
+    final public function addParam(InputParamInterface $param): self
+    {
+        if (! is_array($this->inputParams)) {
+            throw new RuntimeException(sprintf(
+                'Command %s uses $inputParams property; please do not override that property when using %s,'
+                . ' as it becomes incompatible with input parameter usage',
+                static::class,
+                __NAMESPACE__ . '\ParamAwareCommandTrait'
+            ));
+        }
+
+        $name = $param->getName();
+
+        $this->addOption(
+            $name,
+            $param->getShortcut(),
+            $param->getOptionMode(),
+            $param->getDescription()
+            // default null, on purpose
+        );
+
+        $this->inputParams[$name] = $param;
+
+        return $this;
+    }
+
+    final public function run(InputInterface $input, OutputInterface $output): int
```

Comment:
The other way we've tried it was in the `Application` instance (by providing a custom application). The problem with that approach is that if the command is then used in another `Application` type, the decoration does not happen, and calls to `addParam()` fail.

The only other approach I can think of is to create an `AbstractParamAwareCommand` that overrides `run()`, but doesn't make it `final`; this would allow users to override it as needed.

Generally speaking, the public API for command authors is restricted to `configure()` and `execute()`, there's very little reason to override `run()`, unless you want to do something like we're doing here, which is an atypical case at best.

So, the question then needs to be: **do we make this an abstract class instead of a trait?**

@michalbundyra Thoughts?

Reply:
I would probably be +1 on this until there aint be a very good reason not to do it. 

However, one could still implement `run` himself if overriding it during `use` in the class, thus, this wont be a blocker for me.

Reply:
@weierophinney as we discussed before on Slack, I can see some props and cons for both solutions. In general I think if we go with abstract class it's gonna be a bit cleaner solution in general.